File: Level 2/전력망을 둘로 나누기/solution.js
function solution(n, wires) {
  const graph = new Graph(n, wires);

  let answer = Infinity;

  wires.forEach(([a, b]) => {
    graph.cutWire(a, b);

    countA = graph.dfs(a, b);
    countB = n - countA;

    answer = Math.min(answer, Math.abs(countA - countB));

    graph.connectWire(a, b);
  });

  return answer;
}

class Graph {
  constructor(n, wires) {
    this.graph = Array.from({ length: n + 1 }, () => []);

    wires.forEach(([a, b]) => {
      this.graph[a].push(b);
      this.graph[b].push(a);
    });
  }

  cutWire(a, b) {
    this.graph[a] = this.graph[a].filter((item) => item !== b);
    this.graph[b] = this.graph[b].filter((item) => item !== a);
  }

  connectWire(a, b) {
    this.graph[a].push(b);
    this.graph[b].push(a);
  }

  dfs(node, parent) {
    let count = 1;

    this.graph[node].forEach((child) => {
      if (child !== parent) {
        count += this.dfs(child, node);
      }
    });

    return count;
  }
}

File: Level 2/전력망을 둘로 나누기/solution.py
class Graph:
    def __init__(self, n, wires):
        self.graph = [[] for _ in range(n + 1)]

        for a, b in wires:
            self.graph[a].append(b)
            self.graph[b].append(a)

    def cut_wire(self, a, b):
        self.graph[a].remove(b)
        self.graph[b].remove(a)

    def connect_wire(self, a, b):
        self.graph[a].append(b)
        self.graph[b].append(a)

    def dfs(self, node, parent):
        count = 1

        for child in self.graph[node]:
            if child != parent:
                count += self.dfs(child, node)

        return count


def solution(n, wires):
    graph = Graph(n, wires)

    answer = float("inf")

    for a, b in wires:
        graph.cut_wire(a, b)

        count_a = graph.dfs(a, b)
        count_b = n - count_a

        answer = min(answer, abs(count_a - count_b))

        graph.connect_wire(a, b)

    return answer
